src/librustc_mir/interpret/visitor.rs

```diff
@@ -0,0 +1,321 @@
+        field: u64,
+    ) -> EvalResult<'tcx, Self> {
+        ectx.mplace_field(self, field)
+    }
+}
+impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> Value<'a, 'mir, 'tcx, M>
+    for PlaceTy<'tcx, M::PointerTag>
+{
+    #[inline(always)]
+    fn layout(&self) -> TyLayout<'tcx> {
+        self.layout
+    }
+
+    #[inline(always)]
+    fn to_mem_place(
+        self,
+        ectx: &EvalContext<'a, 'mir, 'tcx, M>,
+    ) -> EvalResult<'tcx, MPlaceTy<'tcx, M::PointerTag>> {
+        // If this refers to a local, assert that it already has an allocation.
+        Ok(ectx.place_to_op(self)?.to_mem_place())
+    }
+
+    #[inline(always)]
+    fn from_mem_place(mplace: MPlaceTy<'tcx, M::PointerTag>) -> Self {
+        mplace.into()
+    }
+
+    #[inline(always)]
+    fn project_downcast(
+        self,
+        ectx: &EvalContext<'a, 'mir, 'tcx, M>
+    ) -> EvalResult<'tcx, (Self, usize)> {
+        let idx = ectx.read_discriminant(ectx.place_to_op(self)?)?.1;
+        Ok((ectx.place_downcast(self, idx)?, idx))
+    }
+
+    #[inline(always)]
+    fn project_field(
+        self,
+        ectx: &mut EvalContext<'a, 'mir, 'tcx, M>,
+        field: u64,
+    ) -> EvalResult<'tcx, Self> {
+        ectx.place_field(self, field)
+    }
+}
+
+// How to traverse a value and what to do when we are at the leaves.
+pub trait ValueVisitor<'a, 'mir, 'tcx: 'mir+'a, M: Machine<'a, 'mir, 'tcx>>: Sized {
+    type V: Value<'a, 'mir, 'tcx, M>;
+
+    // The visitor must have an `EvalContext` in it.
+    fn ecx(&mut self) -> &mut EvalContext<'a, 'mir, 'tcx, M>;
+
+    // Recursie actions, ready to be overloaded.
+    /// Visit the given value, dispatching as appropriate to more speicalized visitors.
+    #[inline(always)]
+    fn visit_value(&mut self, v: Self::V) -> EvalResult<'tcx>
+    {
+        self.walk_value(v)
+    }
+    /// Visit the given value as a union.
+    #[inline(always)]
+    fn visit_union(&mut self, _v: Self::V) -> EvalResult<'tcx>
+    {
+        Ok(())
+    }
+    /// Visit the given value as an array.
+    #[inline(always)]
+    fn visit_array(&mut self, v: Self::V) -> EvalResult<'tcx>
+    {
+        self.walk_array(v)
+    }
+    /// Called each time we recurse down to a field, passing in old and new value.
+    /// This gives the visitor the chance to track the stack of nested fields that
+    /// we are descending through.
+    #[inline(always)]
+    fn visit_field(
+        &mut self,
+        _old_val: Self::V,
+        _field: usize,
+        new_val: Self::V,
+    ) -> EvalResult<'tcx> {
+        self.visit_value(new_val)
+    }
+
+    // Actions on the leaves, ready to be overloaded.
+    /// Called whenever we reach a value with uninhabited layout.
+    /// Recursing to fields will continue after this!
+    #[inline(always)]
+    fn visit_uninhabited(&mut self, _v: Self::V) -> EvalResult<'tcx>
+    { Ok(()) }
+    /// Called whenever we reach a value with scalar layout.
+    /// Recursing to fields will continue after this!
```

Comment:
I think the default impl should be calling a `walk_scalar` method then. Pass all arguments needed to recurse. Implementing a `visit_*` function as `Ok(())` should stop recursion right there.

Reply:
That does not match the structure we used for the validation visitor, and I think it is the wrong structure. The point is that being a scalar and being an aggregate are orthogonal properties, a value can be neither or both or any combination. Some scalars are primitives, some are not - -we'd have to duplicate all the primitive-vs-aggregate handling. I don't think that would be a good idea.

Reply:
Oh right. Add a comment that it's not really recursing into the scalar, but the value, so one should overwrite `visit_value` if they want to not recurse into this case

Reply:
Like so?